### DISCUSSION TITLE
Multithreading in Python and ZOAU with mvscmd.execute_authorized

### DISCUSSION THREAD

**Discussion by EnsonoStevenPerva** [IBM/zoau-samples#18] [Q&A] [answered]
I am having trouble with `mvscmd.execute_authorized` blocking other threads in a Python script until the address space returns. Does anyone have any experience with running multiple threads and utilizing ZOAU?

Here is the use-case:
Execute ICKDSF to INIT a volume, ICKDSF issues a WTOR and the address space does not return until a reply to the WTOR is issued.
Spawn a thread before issuing the execute that monitors for outstanding replies and replies to them if they fit appropriate criteria.

Admittedly, this use-case is weak because of the `NOREPLYU,FORCE` parameter, but I feel it is demonstrative of the symptom I'd like to resolve.

I can share a code sample that demonstrates this if necessary. I am using IBM Open Enterprise Python 3.9.2 and ZOAU 1.2.0. If this is a known limitation, I am happy to open an Issue here if that is helpful.

**Comment by MikeFultonDev** [comment]
Interesting question. My first thought is whether this blocking is happening because of something in Python or is in mvscmdauth. mvscmd and mvscmdauth both call BPX1ATM under the covers and it might do this blocking.

**Reply by EnsonoStevenPerva** [reply]
I have been able to successfully do threading in Python outside of `mvscmd.execute_authorized`. I can provide a simple example of threading that is functional from a Python perspective:
```
import os
import threading
from time import sleep

def counter():
    accum = 0
    while accum < 60:
        print(f'Iteration {accum}...')
        accum += 1
        sleep(1)

count_thread = threading.Thread(target=counter)
count_thread.start()

main_accum = 0
while True:
    print(f'Main Count {main_accum}...')
    main_accum += 1
    sleep(0.5)
```
Adding some sort of `mvscmd.execute_authorized` into the mix on either the main thread or the `count_thread` will result in blocking during the call. I suspect it would be the same for a simple `mvscmd.execute` call as well.

**Comment by MikeFultonDev** [comment]
This does appear to be a bug in the ZOAU Python language pack @EnsonoStevenPerva - thanks for raising it!
If you have S&S for ZOAU, please feel free to open a support ticket to expedite a fix.

**Reply by EnsonoStevenPerva** [reply]
Thanks for the follow-through here. I unfortunately do not have S&S for ZOAU. Hopefully this will percolate up the chain somehow and find its way into remediation! I have a few ideas where threading would make life easier...

**Reply by storfibm** [reply]
Thanks for raising this issue. If applicable to your account and if you'd like to engage IBM support, you can open a support case via the following address:
[https://www.ibm.com/mysupport/s/createrecord/NewCase](https://www.ibm.com/mysupport/s/createrecord/NewCase)

Community users also have an ability to collaborate with IBM and make requests for enhancements via the IBM DeveloperNet RFE process at the following address:
[https://www.ibm.com/developerworks/rfe](https://www.ibm.com/developerworks/rfe)

Prioritization of community issues is based on a number of factors, but all requests will be reviewed.

Sebastian Torf
IBM Z Hybrid Cloud Foundation - Product Manager

**Comment by MikeFultonDev** [comment]
Hi

I figured that might be the case. 
Please keep us in the loop and we will also post when we are able to fix this.

**Comment by ddimatos** [comment]
I have captured this into our backlog with JIRA NAZARE-8851. Thank you.

**Comment by storfibm** [comment, marked as answer]
A relevant code sample showcasing Python multiprocessing and how you might be able to achieve this use case with ZOAU has now been supplied in the "/samples" directory of the latest ZOAU 1.2.4 release (PH52034/UI92506)
